automerge/src/autotxn.rs

```diff
@@ -0,0 +1,417 @@
+use crate::exid::ExId;
+use crate::transaction::CommitOptions;
+use crate::types::Patch;
+use crate::{
+    change::export_change, transaction::TransactionInner, ActorId, Automerge, AutomergeError,
+    Change, ChangeHash, Prop, Value,
+};
+use crate::{SyncMessage, SyncState};
+
+/// An automerge document that automatically manages transactions.
+#[derive(Debug, Clone)]
+pub struct AutoTxn {
+    doc: Automerge,
+    transaction: Option<TransactionInner>,
+}
+
+impl Default for AutoTxn {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl AutoTxn {
+    pub fn new() -> Self {
+        Self {
+            doc: Automerge::new(),
+            transaction: None,
+        }
+    }
+
+    /// Get the inner document.
+    #[doc(hidden)]
+    pub fn document(&mut self) -> &Automerge {
+        self.ensure_transaction_closed();
+        &self.doc
+    }
+
+    pub fn set_actor(&mut self, actor: ActorId) {
+        self.ensure_transaction_closed();
+        self.doc.set_actor(actor)
+    }
+
+    pub fn get_actor(&mut self) -> ActorId {
+        self.doc.get_actor()
+    }
+
+    pub fn maybe_get_actor(&self) -> Option<ActorId> {
+        self.doc.maybe_get_actor()
+    }
+
+    pub fn new_with_actor_id(actor: ActorId) -> Self {
+        Self {
+            doc: Automerge::new_with_actor_id(actor),
+            transaction: None,
+        }
+    }
+
+    pub fn pending_ops(&self) -> usize {
+        self.transaction
+            .as_ref()
+            .map(|t| t.pending_ops())
+            .unwrap_or(0)
+    }
+
+    fn try_start_transaction(&mut self) {
```

Comment:
I think a better name for this would be `ensure_transaction_open` for the symmetry with `ensure_transaction_closed`